src/rust/bitbox02-noise/src/noise_xx.rs

```diff
@@ -0,0 +1,297 @@
+            } => Ok(*remote_static_pubkey),
+            _ => Err(Error::WrongState),
+        }
+    }
+
+    /// Mark the pairing as verified, unlocking encryption and decryption.
+    ///
+    /// This should only be called after the user verified the pairing code (see
+    /// `get_handshake_hash()`), or has done so in the past for the same remote communucation
+    /// partner (see `remote_static_pubkey()`).
+    pub fn set_pairing_verified(&mut self) -> Result<(), Error> {
+        match self {
+            State::Ready {
+                pairing_verification_required,
+                ..
+            } => {
+                *pairing_verification_required = false;
+                Ok(())
+            }
+            _ => Err(Error::WrongState),
+        }
+    }
+
+    /// Decrypt an encrypted message.
+    pub fn decrypt(&mut self, msg: &[u8]) -> Result<Vec<u8>, Error> {
+        match self {
+            State::Ready {
+                pairing_verification_required: true,
+                ..
+            } => Err(Error::PairingVerificationRequired),
+
+            State::Ready {
+                pairing_verification_required: false,
+                receive,
+                ..
+            } => match receive.decrypt_vec(&msg) {
+                Ok(r) => Ok(r),
+                Err(()) => Err(Error::Noise),
+            },
+            _ => Err(Error::WrongState),
+        }
+    }
+
+    /// Encrypt a message. The ciphertext is appended to `out`.
+    pub fn encrypt(&mut self, msg: &[u8], out: &mut Vec<u8>) -> Result<(), Error> {
+        match self {
+            State::Ready {
+                pairing_verification_required: true,
+                ..
+            } => Err(Error::PairingVerificationRequired),
+
+            State::Ready {
+                pairing_verification_required: false,
+                send,
+                ..
+            } => {
+                let start = out.len();
+                let encrypted_len = msg.len() + 16;
+                // Make space for result.
+                out.resize(start + encrypted_len, 0);
+                send.encrypt(&msg, &mut out[start..]);
+                Ok(())
+            }
+            _ => Err(Error::WrongState),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    impl HandshakeResult {
+        fn response(self) -> Result<Vec<u8>, ()> {
+            match self {
+                HandshakeResult::Response(r) => Ok(r),
+                HandshakeResult::Done => Err(()),
+            }
+        }
+        fn done(self) -> Result<(), ()> {
+            match self {
+                HandshakeResult::Response(_) => Err(()),
+                HandshakeResult::Done => Ok(()),
+            }
+        }
+    }
+
+    enum MockRandom32 {}
+    impl Random32 for MockRandom32 {
+        fn mcu_32_bytes(out: &mut [u8; 32]) {
+            out.copy_from_slice(b"llllllllllllllllllllllllllllllll")
+        }
+    }
+
+    #[test]
+    pub fn test_full() {
```

Comment:
will add more complete test coverage (negative tests) if it looks like the code won't change much anymore.